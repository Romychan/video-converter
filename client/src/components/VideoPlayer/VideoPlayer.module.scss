.container {
  width: 100%;
}

.video {
  width: 100%;
  max-height: 410px;
  background: var(--color-primary);
  border-radius: 10px;
}

.controls {
}
