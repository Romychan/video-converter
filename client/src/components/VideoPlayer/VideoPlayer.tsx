import {
  ChangeEvent,
  MutableRefObject,
  useEffect,
  useRef,
  useState,
} from 'react';

import { toHHMMSS } from '@utils/toHHMMSS';

import { Actions } from './Actions/Actions';
import { TimeLine } from './TimeLine/TimeLine';

import styles from './VideoPlayer.module.scss';

interface IVideoPlayerProps {
  uri: string;
  handleLoadedVideo: (value: string) => void;
}

export const VideoPlayer = ({ uri, handleLoadedVideo }: IVideoPlayerProps) => {
  const videoElement = useRef() as MutableRefObject<HTMLVideoElement>;
  const [duration, setDuration] = useState(0);
  const [playerState, setPlayerState] = useState({
    isPlaying: false,
    progress: 0,
  });

  const togglePlay = () => {
    setPlayerState({
      ...playerState,
      isPlaying: !playerState.isPlaying,
    });
  };

  const handleOnTimeUpdate = () => {
    const progress =
      (videoElement.current.currentTime / videoElement.current.duration) * 100;

    setPlayerState({
      ...playerState,
      progress,
    });
  };

  const handleVideoProgress = (event: ChangeEvent<HTMLInputElement>) => {
    const manualChange = parseInt(event.target.value);
    videoElement.current.currentTime =
      (videoElement.current.duration / 100) * manualChange;

    setPlayerState({
      ...playerState,
      progress: manualChange,
    });
  };

  const handleLoadedMetadata = () => {
    const video = videoElement.current;

    if (!video) return;

    setDuration(video.duration);
    handleLoadedVideo(toHHMMSS(video.duration));
  };

  useEffect(() => {
    playerState.isPlaying
      ? videoElement.current.play()
      : videoElement.current.pause();
  }, [playerState.isPlaying, videoElement]);

  return (
    <div className={styles.container}>
      <video
        className={styles.video}
        src={uri}
        ref={videoElement}
        onTimeUpdate={handleOnTimeUpdate}
        onLoadedMetadata={handleLoadedMetadata}
        data-testid="video-player"
      />

      <div className={styles.controls}>
        <TimeLine
          progress={playerState.progress}
          handleVideoProgress={handleVideoProgress}
        />

        <Actions
          isPlaying={playerState.isPlaying}
          progress={playerState.progress}
          duration={duration}
          togglePlay={togglePlay}
        />
      </div>
    </div>
  );
};
