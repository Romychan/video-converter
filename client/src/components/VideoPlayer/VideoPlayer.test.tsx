/* eslint-disable @typescript-eslint/no-empty-function */
import {
  fireEvent,
  screen,
  render,
  act,
  RenderResult,
} from '@testing-library/react';
import { vi } from 'vitest';

import { VideoPlayer } from './VideoPlayer';

describe('TimeLine', () => {
  const handleChange = vi.fn();
  let container: RenderResult;

  beforeEach(async () => {
    container = render(
      <VideoPlayer uri="test" handleLoadedVideo={handleChange} />,
    );

    const inlineVideo = screen.getByTestId('video-player');
    Object.defineProperty(inlineVideo, 'duration', {
      writable: true,
      value: 60,
    });

    await act(async () => fireEvent.loadedMetadata(inlineVideo));
  });

  it('should render correctly', async () => {
    expect(screen.getByTestId('video-player')).toHaveProperty('duration', 60);
    expect(screen.getByText(/00:01:00/i)).toBeInTheDocument();
    expect(handleChange).toHaveBeenCalled();
  });

  it('should change time', async () => {
    const timelineElement = screen.getByLabelText('Timeline');

    fireEvent.change(timelineElement, { target: { value: '10' } });
    expect(timelineElement).toHaveValue('10');
    expect(screen.getByText(/00:00:06/i)).toBeInTheDocument();
  });

  it('should play video', async () => {
    const playSub = vi
      .spyOn(window.HTMLMediaElement.prototype, 'play')
      .mockImplementation((): any => {});
    const buttonElement = screen.getByTestId('actions-button');
    fireEvent.click(buttonElement);

    expect(playSub).toHaveBeenCalled();
    playSub.mockRestore();
  });

  it('should match snapshot', () => {
    expect(container).toMatchSnapshot();
  });
});
