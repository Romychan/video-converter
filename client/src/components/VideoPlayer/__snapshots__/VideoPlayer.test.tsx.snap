// Vitest Snapshot v1

exports[`TimeLine > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_443994"
      >
        <video
          class="_video_443994"
          data-testid="video-player"
          src="test"
        />
        <div
          class="_controls_443994"
        >
          <input
            aria-label="Timeline"
            class="_timeline_2c5235"
            max="100"
            min="0"
            type="range"
            value="0"
          />
          <div
            class="_actions_449f60"
          >
            <button
              class="_square_449f60"
              data-testid="actions-button"
            >
              <div
                class="_playback_449f60"
                data-testid="actions-playing"
              />
            </button>
            <p
              class="text text_small"
            >
              00:00:00
               / 
              00:01:00
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_container_443994"
    >
      <video
        class="_video_443994"
        data-testid="video-player"
        src="test"
      />
      <div
        class="_controls_443994"
      >
        <input
          aria-label="Timeline"
          class="_timeline_2c5235"
          max="100"
          min="0"
          type="range"
          value="0"
        />
        <div
          class="_actions_449f60"
        >
          <button
            class="_square_449f60"
            data-testid="actions-button"
          >
            <div
              class="_playback_449f60"
              data-testid="actions-playing"
            />
          </button>
          <p
            class="text text_small"
          >
            00:00:00
             / 
            00:01:00
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
